File: backend/models/category.go
package models

import (
	"strings"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Category struct {
	ID          primitive.ObjectID  `bson:"_id,omitempty" json:"id"`
	Name        string              `bson:"name" json:"name"`
	Description string              `bson:"description" json:"description"`
	ImageURL    string              `bson:"imageUrl" json:"imageUrl"`
	ParentID    *primitive.ObjectID `bson:"parentId,omitempty" json:"parentId"`
	Path        string              `bson:"path" json:"path"`
	Depth       int                 `bson:"depth" json:"depth"`
	Children    []*Category         `bson:"-" json:"children,omitempty"`
	CreatedAt   time.Time           `bson:"createdAt" json:"createdAt"`
	UpdatedAt   time.Time           `bson:"updatedAt" json:"updatedAt"`
}

func (c *Category) SetPath(parentPath string) {
	if parentPath == "" {
		c.Path = "/" + c.ID.Hex() + "/"
		c.Depth = 1
	} else {
		c.Path = parentPath + c.ID.Hex() + "/"
		c.Depth = strings.Count(parentPath, "/")
	}
}

func (c *Category) IsDescendantOf(parentID primitive.ObjectID) bool {
	return strings.Contains(c.Path, "/"+parentID.Hex()+"/")
}

func (c *Category) IsRoot() bool {
	return c.ParentID == nil
}

File: admin_console/lib/models/category.dart
class Category {
  final String id;
  final String name;
  final String? description;
  final String? imageUrl;
  final String? parentId;
  final String path;
  final int depth;
  final List<Category> children;

  Category({
    required this.id,
    required this.name,
    this.description,
    this.imageUrl,
    this.parentId,
    required this.path,
    required this.depth,
    this.children = const [],
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      imageUrl: json['imageUrl'],
      parentId: json['parentId'],
      path: json['path'],
      depth: json['depth'],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'description': description,
    'imageUrl': imageUrl,
    'parentId': parentId,
    'path': path,
    'depth': depth,
  };

  bool get isRoot => parentId == null;

  Category copyWith({
    String? id,
    String? name,
    String? description,
    String? imageUrl,
    String? parentId,
    String? path,
    int? depth,
    List<Category>? children,
  }) {
    return Category(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      imageUrl: imageUrl ?? this.imageUrl,
      parentId: parentId ?? this.parentId,
      path: path ?? this.path,
      depth: depth ?? this.depth,
      children: children ?? this.children,
    );
  }
}